### PR TITLE
Make `experimental.serverSourceMaps` option to stable

### PR DESCRIPTION
**Pull request by devjiwonchoi**
### Why?

Using source maps can be critical for debugging, and an option to enable it should not stay experimental.

Closes NDX-427

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js jwn/32iu | Change |
| - | - | - | - |
| buildDuration | 20.2s | 18.2s | N/A |
| buildDurationCached | 17.5s | 15.3s | N/A |
| nodeModulesSize | 403 MB | 403 MB | N/A |
| nextStartRea..uration (ms) | 456ms | 454ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js jwn/32iu | Change |
| - | - | - | - |
| 4789-HASH.js gzip | 5.27 kB | 5.27 kB | N/A |
| 8781.HASH.js gzip | 169 B | 169 B | ✓ |
| 9785-HASH.js gzip | 45.4 kB | 45.4 kB | N/A |
| b5cfd723-HASH.js gzip | 52.5 kB | 52.5 kB | N/A |
| framework-HASH.js gzip | 57.3 kB | 57.3 kB | N/A |
| main-app-HASH.js gzip | 233 B | 231 B | N/A |
| main-HASH.js gzip | 33.1 kB | 33.1 kB | N/A |
| webpack-HASH.js gzip | 1.71 kB | 1.71 kB | N/A |
| Overall change | 169 B | 169 B | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js jwn/32iu | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 39.4 kB | 39.4 kB | ✓ |
| Overall change | 39.4 kB | 39.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js jwn/32iu | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 193 B | N/A |
| _error-HASH.js gzip | 193 B | 192 B | N/A |
| amp-HASH.js gzip | 511 B | 508 B | N/A |
| css-HASH.js gzip | 343 B | 343 B | ✓ |
| dynamic-HASH.js gzip | 1.84 kB | 1.84 kB | N/A |
| edge-ssr-HASH.js gzip | 266 B | 264 B | N/A |
| head-HASH.js gzip | 362 B | 365 B | N/A |
| hooks-HASH.js gzip | 392 B | 391 B | N/A |
| image-HASH.js gzip | 4.41 kB | 4.41 kB | N/A |
| index-HASH.js gzip | 268 B | 268 B | ✓ |
| link-HASH.js gzip | 2.78 kB | 2.77 kB | N/A |
| routerDirect..HASH.js gzip | 329 B | 327 B | N/A |
| script-HASH.js gzip | 397 B | 394 B | N/A |
| withRouter-HASH.js gzip | 324 B | 322 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 717 B | 717 B | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js jwn/32iu | Change |
| - | - | - | - |
| _buildManifest.js gzip | 751 B | 752 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js jwn/32iu | Change |
| - | - | - | - |
| index.html gzip | 523 B | 524 B | N/A |
| link.html gzip | 538 B | 539 B | N/A |
| withRouter.html gzip | 519 B | 519 B | ✓ |
| Overall change | 519 B | 519 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js jwn/32iu | Change |
| - | - | - | - |
| edge-ssr.js gzip | 128 kB | 128 kB | N/A |
| page.js gzip | 189 kB | 189 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js jwn/32iu | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 671 B | 667 B | N/A |
| middleware-r..fest.js gzip | 155 B | 155 B | ✓ |
| middleware.js gzip | 31 kB | 31 kB | N/A |
| edge-runtime..pack.js gzip | 844 B | 844 B | ✓ |
| Overall change | 999 B | 999 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js jwn/32iu | Change |
| - | - | - | - |
| 973-experime...dev.js gzip | 322 B | 322 B | ✓ |
| 973.runtime.dev.js gzip | 314 B | 314 B | ✓ |
| app-page-exp...dev.js gzip | 318 kB | 318 kB | ✓ |
| app-page-exp..prod.js gzip | 122 kB | 122 kB | ✓ |
| app-page-tur..prod.js gzip | 135 kB | 135 kB | ✓ |
| app-page-tur..prod.js gzip | 130 kB | 130 kB | ✓ |
| app-page.run...dev.js gzip | 309 kB | 309 kB | ✓ |
| app-page.run..prod.js gzip | 118 kB | 118 kB | ✓ |
| app-route-ex...dev.js gzip | 35.9 kB | 35.9 kB | ✓ |
| app-route-ex..prod.js gzip | 24.3 kB | 24.3 kB | ✓ |
| app-route-tu..prod.js gzip | 24.4 kB | 24.4 kB | ✓ |
| app-route-tu..prod.js gzip | 24.1 kB | 24.1 kB | ✓ |
| app-route.ru...dev.js gzip | 37.5 kB | 37.5 kB | ✓ |
| app-route.ru..prod.js gzip | 24.1 kB | 24.1 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.59 kB | 9.59 kB | ✓ |
| pages-api.ru...dev.js gzip | 11.4 kB | 11.4 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.58 kB | 9.58 kB | ✓ |
| pages-turbo...prod.js gzip | 20.9 kB | 20.9 kB | ✓ |
| pages.runtim...dev.js gzip | 26.5 kB | 26.5 kB | ✓ |
| pages.runtim..prod.js gzip | 20.9 kB | 20.9 kB | ✓ |
| server.runti..prod.js gzip | 914 kB | 914 kB | ✓ |
| Overall change | 2.32 MB | 2.32 MB | ✓ |

</details>

<details>
<summary><strong>build cache</strong></summary>

|  | vercel/next.js canary  | vercel/next.js jwn/32iu | Change |
| - | - | - | - |
| 0.pack gzip | 1.86 MB | 1.86 MB | N/A |
| index.pack gzip | 144 kB | 143 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>image-HASH.js</strong></summary>

```diff
@@ -1,7 +1,7 @@
 (self["webpackChunk_N_E"] = self["webpackChunk_N_E"] || []).push([
   [8358],
   {
-    /***/ 869: /***/ (
+    /***/ 8942: /***/ (
       __unused_webpack_module,
       __unused_webpack_exports,
       __webpack_require__
@@ -9,7 +9,7 @@
       (window.__NEXT_P = window.__NEXT_P || []).push([
         "/image",
         function () {
-          return __webpack_require__(8560);
+          return __webpack_require__(2864);
         },
       ]);
       if (false) {
@@ -18,7 +18,7 @@
       /***/
     },
 
-    /***/ 5689: /***/ (module, exports, __webpack_require__) => {
+    /***/ 1673: /***/ (module, exports, __webpack_require__) => {
       "use strict";
       /* __next_internal_client_entry_do_not_use__  cjs */
       Object.defineProperty(exports, "__esModule", {
@@ -40,17 +40,17 @@
         __webpack_require__(8693)
       );
       const _head = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(6998)
+        __webpack_require__(5838)
       );
-      const _getimgprops = __webpack_require__(5861);
-      const _imageconfig = __webpack_require__(1253);
-      const _imageconfigcontextsharedruntime = __webpack_require__(6305);
-      const _warnonce = __webpack_require__(1629);
-      const _routercontextsharedruntime = __webpack_require__(1616);
+      const _getimgprops = __webpack_require__(5726);
+      const _imageconfig = __webpack_require__(187);
+      const _imageconfigcontextsharedruntime = __webpack_require__(5786);
+      const _warnonce = __webpack_require__(3673);
+      const _routercontextsharedruntime = __webpack_require__(759);
       const _imageloader = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(7106)
+        __webpack_require__(2182)
       );
-      const _usemergedref = __webpack_require__(9916);
+      const _usemergedref = __webpack_require__(9363);
       // This is replaced by webpack define plugin
       const configEnv = {
         deviceSizes: [640, 750, 828, 1080, 1200, 1920, 2048, 3840],
@@ -371,7 +371,7 @@
       /***/
     },
 
-    /***/ 9916: /***/ (module, exports, __webpack_require__) => {
+    /***/ 9363: /***/ (module, exports, __webpack_require__) => {
       "use strict";
 
       Object.defineProperty(exports, "__esModule", {
@@ -432,7 +432,7 @@
       /***/
     },
 
-    /***/ 5861: /***/ (
+    /***/ 5726: /***/ (
       __unused_webpack_module,
       exports,
       __webpack_require__
@@ -448,9 +448,9 @@
           return getImgProps;
         },
       });
-      const _warnonce = __webpack_require__(1629);
-      const _imageblursvg = __webpack_require__(2040);
-      const _imageconfig = __webpack_require__(1253);
+      const _warnonce = __webpack_require__(3673);
+      const _imageblursvg = __webpack_require__(5499);
+      const _imageconfig = __webpack_require__(187);
       const VALID_LOADING_VALUES =
         /* unused pure expression or super */ null && [
           "lazy",
@@ -824,7 +824,7 @@
       /***/
     },
 
-    /***/ 2040: /***/ (__unused_webpack_module, exports) => {
+    /***/ 5499: /***/ (__unused_webpack_module, exports) => {
       "use strict";
       /**
        * A shared function, used on both client and server, to generate a SVG blur placeholder.
@@ -879,7 +879,7 @@
       /***/
     },
 
-    /***/ 4994: /***/ (
+    /***/ 6157: /***/ (
       __unused_webpack_module,
       exports,
       __webpack_require__
@@ -906,10 +906,10 @@
         },
       });
       const _interop_require_default = __webpack_require__(9608);
-      const _getimgprops = __webpack_require__(5861);
-      const _imagecomponent = __webpack_require__(5689);
+      const _getimgprops = __webpack_require__(5726);
+      const _imagecomponent = __webpack_require__(1673);
       const _imageloader = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(7106)
+        __webpack_require__(2182)
       );
       function getImageProps(imgProps) {
         const { props } = (0, _getimgprops.getImgProps)(imgProps, {
@@ -941,7 +941,7 @@
       /***/
     },
 
-    /***/ 7106: /***/ (__unused_webpack_module, exports) => {
+    /***/ 2182: /***/ (__unused_webpack_module, exports) => {
       "use strict";
 
       Object.defineProperty(exports, "__esModule", {
@@ -976,7 +976,7 @@
       /***/
     },
 
-    /***/ 8560: /***/ (
+    /***/ 2864: /***/ (
       __unused_webpack_module,
       __webpack_exports__,
       __webpack_require__
@@ -993,8 +993,8 @@
 
       // EXTERNAL MODULE: ./node_modules/.pnpm/react@19.0.0-rc-7c8e5e7a-20241101/node_modules/react/jsx-runtime.js
       var jsx_runtime = __webpack_require__(2073);
-      // EXTERNAL MODULE: ./node_modules/.pnpm/next@file+..+main-repo+packages+next+next-packed.tgz_react-dom@19.0.0-rc-7c8e5e7a-20241101_re_ngsqfltlxii45umrwe6rktuauy/node_modules/next/image.js
-      var next_image = __webpack_require__(4678);
+      // EXTERNAL MODULE: ./node_modules/.pnpm/next@file+..+diff-repo+packages+next+next-packed.tgz_react-dom@19.0.0-rc-7c8e5e7a-20241101_re_ay5jspqvpcpk5357qiimbatlya/node_modules/next/image.js
+      var next_image = __webpack_require__(9915);
       var image_default = /*#__PURE__*/ __webpack_require__.n(next_image); // CONCATENATED MODULE: ./pages/nextjs.png
       /* harmony default export */ const nextjs = {
         src: "/_next/static/media/nextjs.cae0b805.png",
@@ -1024,12 +1024,12 @@
       /***/
     },
 
-    /***/ 4678: /***/ (
+    /***/ 9915: /***/ (
       module,
       __unused_webpack_exports,
       __webpack_require__
     ) => {
-      module.exports = __webpack_require__(4994);
+      module.exports = __webpack_require__(6157);
 
       /***/
     },
@@ -1039,7 +1039,7 @@
     /******/ var __webpack_exec__ = (moduleId) =>
       __webpack_require__((__webpack_require__.s = moduleId));
     /******/ __webpack_require__.O(0, [2888, 9774, 179], () =>
-      __webpack_exec__(869)
+      __webpack_exec__(8942)
     );
     /******/ var __webpack_exports__ = __webpack_require__.O();
     /******/ _N_E = __webpack_exports__;
```
</details>

</details>

</details>
Commit: d57d02eb9ece7cb1c162f62fc910f305e3888d9b

**Comment by eps1lon** [comment]
What was the success criteria for this experiment initial? For example, did we want to stress test something or get some stakeholder feedback? I'd check with the person who originally added this experiment.

**Comment by eps1lon** [comment]
We should also discuss how these interact with builds and deploys. I'd like to just enable them by default in build but haven't checked the security implications if this means we also push them to the hosting provider.

**Comment by eps1lon** [comment]
There's also no documentation for this feature. The only mention is https://nextjs.org/docs/app/building-your-application/optimizing/memory-usage which is incorrect because `next build` already defaults to `experimental.serverSourceMaps` according to  @unstubbable. Only `next dev` have them by default.

But we should also explain what happens if you don't have sourcemaps for server code with regards to debugging errors.

**Comment by unstubbable** [comment]
> [...] `next build` already defaults to `experimental.serverSourceMaps` according to @unstubbable. Only `next dev` have them by default.

`next build` does not default to generating server source maps, unless it's for the edge runtime, otherwise only `next dev` does. See https://github.com/vercel/next.js/blob/a5c7884921c190fe69ed30c1f401272cc9632ebe/packages/next/src/build/webpack/config/blocks/base.ts#L34-L63

**Review by ztanner — COMMENTED** [review]
We should probably treat this the same as other features that move out of experimental and add a warning, eg: https://github.com/vercel/next.js/blob/9bbf1d023cc8ae9a371c68c169869a74b39514d3/packages/next/src/server/config.ts#L464-L470

File: docs/02-app/02-api-reference/05-next-config-js/serverSourceMaps.mdx [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+/** @type {import('next').NextConfig} */
+
+const nextConfig = {
+  serverSourceMaps: true,
+}
+
+module.exports = nextConfig
+```
+
+```js filename="next.config.mjs" switcher
```
Comment (delbaoliveira) [inline docs/02-app/02-api-reference/05-next-config-js/serverSourceMaps.mdx:18]: I'm not sure 3 code blocks will work as we only support JS and TS. Maybe you can update the filename for the first one to be filename="next.config.js|mjs"?

**Review by delbaoliveira — CHANGES_REQUESTED** [review]
Thank you for the docs @devjiwonchoi! 

Is this feature available in pages router or only app router? Do we need to share the page between the docs?

**Review by sokra — COMMENTED** [review]
It should get some tests when it's no longer experimental

**Comment by devjiwonchoi** [comment]
Closing as we need more backup data to make it stable.

**Pull request closed without merging**